feat: headless tree integration

lukasbach opened the pull request:
This PR replaces the React Arborist Setup of the sidepanel trees with Headless Tree.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Switched page and shared trees to a headless, item-centric tree with async on-demand loading, improved drag-and-drop, keyboard support, and added a cached page fetch endpoint.

* **Bug Fixes**
  * Breadcrumbs now show a localized "untitled" label for unnamed pages.

* **Refactor**
  * Migrated tree logic to item-based mutations/hooks, removed legacy tree integration, and updated realtime handling to use tree-instance cache updates/invalidation.

* **Style**
  * Updated tree/node styling for drag previews, selection, and a visible drag line.

* **Chores**
  * Tooling/dependency config updates and tightened .vscode ignore.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/docmost/docmost?pullRequest=1457) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/docmost/docmost?pullRequest=1457) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/docmost/docmost?pullRequest=1457) it.</sub>

lukasbach commented on apps/client/src/features/page/components/breadcrumbs/breadcrumb.tsx:
```diff
@@ -39,11 +40,11 @@ export default function Breadcrumb() {
-    if (treeData?.length > 0 && currentPage) {
-      const breadcrumb = findBreadcrumbPath(treeData, currentPage.id);
+    if (treeAtom?.tree?.getItems().length > 0 && currentPage) {
+      const breadcrumb =  findBreadcrumbPathNew(treeAtom.tree.getItemInstance(currentPage.id));
```
Fixed

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -73,195 +50,119 @@ import { mobileSidebarAtom } from "@/components/layouts/global/hooks/atoms/sideb
 import { duplicatePage } from "../../services/page-service.ts";
+import { useTree as useHeadlessTree } from "@headless-tree/react/react17"
+import { asyncDataLoaderFeature, createOnDropHandler, dragAndDropFeature, hotkeysCoreFeature, selectionFeature, type FeatureImplementation, type ItemInstance } from "@headless-tree/core";
+import { t } from "i18next";
```
Fixed

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -73,195 +50,119 @@ import { mobileSidebarAtom } from "@/components/layouts/global/hooks/atoms/sideb
+     onDrop: treeMutations.move,
+     dataLoader: {
+       getItem: async pageId => {
+         return null as any;
+       },
```
Added a comment on the reasoning for this decision.

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -73,195 +50,119 @@ import { mobileSidebarAtom } from "@/components/layouts/global/hooks/atoms/sideb
-
-    fetchData();
-  }, [isDataLoaded, currentPage?.id]);
+    // @ts-ignore
+    setTree({ tree });
+  }, [tree, setTree]);
```
Required due to type limitation of Jotai, this workaround was used before already.

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -73,195 +50,119 @@ import { mobileSidebarAtom } from "@/components/layouts/global/hooks/atoms/sideb
-  }, [setTreeApi]);
+      setTree({ tree }); // trigger rerender of breadcrumbs
+    })();
+  }, [currentPage?.id]);
```
Fixed

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -285,96 +186,67 @@ function Node({ node, style, dragHandle, tree }: NodeRendererProps<any>) {
     updatePageMutation
-      .mutateAsync({ pageId: node.id, icon: emoji.native })
+      .mutateAsync({ pageId: item.getId(), icon: emoji.native })
       .then((data) => {
         setTimeout(() => {
           emit({
             operation: "updateOne",
-            spaceId: node.data.spaceId,
+            spaceId: item.getItemData().spaceId,
             entity: ["pages"],
-            id: node.id,
+            id: item.getId(),
             payload: { icon: emoji.native, parentPageId: data.parentPageId },
           });
+        item.updateCachedData({
+          ...item.getItemData(),
+          icon: emoji.native,
+        });
         }, 50);
       });
```
Timeout was used before already

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -285,96 +186,67 @@ function Node({ node, style, dragHandle, tree }: NodeRendererProps<any>) {
-    }, 650);
+  if (!item.getItemData()) {
+    console.warn("Item data is missing for item:", item.getId());
   }
```
Removed, that was a remainder from experimenting.

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -419,23 +290,15 @@ function Node({ node, style, dragHandle, tree }: NodeRendererProps<any>) {
-    }
+function CreateNode({ item, spaceId }: CreateNodeProps) {
+  const treeMutations = useTreeMutation<SpaceTreeNode>(spaceId);
+  async function handleCreate() {
+    await treeMutations.create(item);
+    item.expand();
   }
```
Throwing upwards should be fine here.

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -666,40 +515,42 @@ function NodeMenu({ node, treeApi, spaceId }: NodeMenuProps) {
   useEffect(() => {
-    if (node.isOpen) {
-      onExpandTree();
-    }
-  }, []);
+    // isFolder() will dispatch retrieval of children if not loaded,
+    // so let's not call that during render
+    setIsFolder(item.isFolder());
+  });
```
That's intentional, isFolder should run on a subsequent render and is very fast. `item` on the other hand is a stable reference and cannot be used for determining when to call this.

lukasbach commented on apps/client/src/features/page/tree/hooks/use-tree-mutation.ts:
```diff
@@ -142,115 +91,87 @@ export function useTreeMutation<T>(spaceId: string) {
+      // await Promise.all(items.map(item => item.getParent()?.invalidateChildrenIds()));
+      // await target.item.invalidateChildrenIds();
+
+      if (!target.item.isExpanded()) setTimeout(target.item.expand);
```
Fixed

lukasbach commented on apps/client/src/features/page/tree/hooks/use-tree-mutation.ts:
```diff
@@ -142,115 +91,87 @@ export function useTreeMutation<T>(spaceId: string) {
-          operation: "deleteTreeNode",
-          spaceId: spaceId,
-          payload: { node: node.data },
-        });
+        items.map(item => item.getItemData())
+          .filter(Boolean)
+          .forEach(node => emit({
+            operation: "deleteTreeNode",
+            spaceId: spaceId,
+            payload: { node },
+          }));
       }, 50);
```
Timeout was used before already

lukasbach commented on apps/client/src/features/share/components/shared-tree.tsx:
```diff
@@ -157,17 +180,25 @@ function Node({ node, style, tree }: NodeRendererProps<any>) {
+function PageArrow({ item }: PageArrowProps) {
+  const [isFolder, setIsFolder] = useState(false);
+
+  useEffect(() => {
+    // isFolder() will dispatch retrieval of children if not loaded,
+    // so let's not call that during render
+    setIsFolder(item.isFolder());
+  });
```
See above

Philipinho commented:
The drag overlay should have less opacity. We should hide the action buttons too. This would bring it in line with the current behavior.

https://github.com/user-attachments/assets/3ecce272-feea-414c-bee1-1ea720b257a9

Current UI:

https://github.com/user-attachments/assets/5977e9a0-a525-4664-9487-5f5e361a8541

Philipinho commented:
Here is a reproduction of the tree toggle state bug.

To replicate, create a nested page to three levels and expand them all.
Try to collapse one level, it collapses and auto-expands immediately.

Video:
https://github.com/user-attachments/assets/a180b3c9-f208-45fc-a02b-07e32c55b117

Philipinho commented:
## Scroll
- The page tree does not scroll for tree items beyond the view port.
- We should also auto-scroll to the opened page's tree location. Similar to the current behavior.

## Duplicating page
Duplicating a page refetches the sidebar-pages. We should only update the cache without making calls to the sidebar-pages endpoint.

## Deleting page
Deleting a page fetches the sidebar pages. We should only update the cache without making calls to the sidebar-pages endpoint.

lukasbach commented on apps/client/src/features/page/tree/hooks/use-tree-mutation.ts:
```diff
@@ -142,115 +102,116 @@ export function useTreeMutation<T>(spaceId: string) {
+
+      if (navigateItem) {
+        navigate(getSpaceUrl(navigateItem.getItemData().slugId));
       }
```
Removed, that section was obsolete.

lukasbach commented on apps/client/src/features/page/tree/hooks/use-tree-mutation.ts:
```diff
@@ -142,115 +102,116 @@ export function useTreeMutation<T>(spaceId: string) {
+    if (!node.isFolder()) return false;
+    const children = (tree.retrieveChildrenIds as any)(node.getId(), true); // TODO update type after HT library update
+    return children.some(child => isPageInNode(pageSlug, tree.getItemInstance(child)));
   };
```
Updated

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -514,13 +391,14 @@ function NodeMenu({ node, treeApi, spaceId }: NodeMenuProps) {
+        ...oldChildren.slice(0, newIndex),
+        treeNodeData.id,
+        ...oldChildren.slice(newIndex),
+      ]);
```
Updated

lukasbach commented on apps/client/src/features/page/tree/components/space-tree.tsx:
```diff
@@ -73,195 +47,146 @@ import { mobileSidebarAtom } from "@/components/layouts/global/hooks/atoms/sideb
-    }
-  }, [currentPage?.id]);
+    tree.getActiveItem()?.scrollTo();
+  }, [tree.getActiveItem().getId()]);
```
Updated

lukasbach commented:
@Philipinho I've updated the implementation and added several fixes:

- Sidebar scrolling now works
- Sidebar auto-scrolls to selected item on page load
- Duplicating and deleting items will now updating the cache instead of triggering a refetch, among other parts
- Generally updated the way how items are fetched, the fetch behavior should now be consistent with how it was before
- Fixed the re-expand bug when collapsing folders
- Updated the way how drag overlay is rendered

Let me know if the current implementation works for you!

Philipinho commented:
Thanks @lukasbach, the reported issues are fixed.
I have three observations:
1. You can't drag and drop above the first node.
2. It scrolls unnecessarily when you click on a node already in view (should not scroll).
3. It would be nice to not to highlight parent nodes (when you try to drop a new child) since the highlight flickers.

I will upload video records to make these more clear.

lukasbach commented:
@Philipinho thanks for the feedback, I've updated the PR.
